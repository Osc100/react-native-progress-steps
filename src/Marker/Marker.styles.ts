import { StyleSheet } from 'react-native';

const circleSize = 40;
const circleBorderWidth = 4;
const barWidth = circleBorderWidth;

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  circle: {
    position: 'relative',
    alignItems: 'center',
    justifyContent: 'center',
    borderWidth: circleBorderWidth,
    borderRadius: circleSize / 2,
    borderColor: 'gray',
    width: circleSize,
    height: circleSize,
    fontSize: 1,
  },
  circleInner: {
    position: 'absolute',
    borderRadius: circleSize / 2,
    backgroundColor: 'black',
    width: circleSize,
    height: circleSize,
  },
  circleText: {
    position: 'relative',
    zIndex: 1,
    fontWeight: 'bold',
  },
  bar: {
    position: 'relative',
    flex: 1,
    alignSelf: 'center',
    backgroundColor: 'gray',
    width: barWidth,
  },
  barInner: {
    backgroundColor: 'black',
    width: barWidth,
    height: '100%',
  },
});

export default styles;
