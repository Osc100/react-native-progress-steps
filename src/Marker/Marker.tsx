import { Animated, View } from 'react-native';
import { DEFAULT_COLORS, EASING } from '../constants';
import React, { FC, memo, useEffect, useRef } from 'react';

import type { IMarker } from './Marker.types';
import styles from './Marker.styles';

const Marker: FC<IMarker> = ({
  stepState: {
    stepIndex = 0,
    isCompleted,
    isActive,
    isLast,
    isFirstInteraction,
  } = {},
  colors: {
    text = DEFAULT_COLORS.marker.text,
    line = DEFAULT_COLORS.marker.line,
  } = {},
}) => {
  const scaleAnimation = useRef(new Animated.Value(0)).current;
  const borderColorAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    Animated.parallel(
      [
        Animated.timing(isCompleted ? scaleAnimation : heightAnimation, {
          toValue: isCompleted ? 1 : 0,
          duration: 300,
          easing: EASING,
          useNativeDriver: false,
        }),
        Animated.timing(isCompleted ? heightAnimation : scaleAnimation, {
          toValue: isCompleted ? 1 : 0,
          duration: 300,
          delay: 150,
          easing: EASING,
          useNativeDriver: false,
        }),
      ],
      { stopTogether: false }
    ).start();
  }, [isCompleted]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    Animated.timing(borderColorAnimation, {
      toValue: isActive ? 1 : 0,
      duration: 400,
      easing: EASING,
      useNativeDriver: false,
    }).start();
  }, [isActive]); // eslint-disable-line react-hooks/exhaustive-deps

  const borderColor =
    isFirstInteraction && isActive
      ? line.completed
      : borderColorAnimation.interpolate({
          inputRange: [0, 1],
          outputRange: [line.normal, line.completed],
        });

  const getColor = () => {
    if (isCompleted) {
      return text.completed;
    } else if (isActive) {
      return text.active;
    } else {
      return text.normal;
    }
  };

  const scale = isFirstInteraction && isCompleted ? 1 : scaleAnimation;

  const height =
    isFirstInteraction && isCompleted
      ? '100%'
      : heightAnimation.interpolate({
          inputRange: [0, 1],
          outputRange: ['0%', '100%'],
        });

  return (
    <View style={styles.container}>
      <Animated.View style={[styles.circle, { borderColor }]}>
        <Animated.Text style={[styles.circleText, { color: getColor() }]}>
          {stepIndex + 1}
        </Animated.Text>
        <Animated.View
          style={[
            styles.circleInner,
            { backgroundColor: line.active, transform: [{ scale }] },
          ]}
        />
      </Animated.View>
      {!isLast && (
        <View style={[styles.bar, { backgroundColor: line.normal }]}>
          <Animated.View
            style={[styles.barInner, { backgroundColor: line.active, height }]}
          />
        </View>
      )}
    </View>
  );
};

export default memo(
  Marker,
  (prevProps, nextProps) =>
    prevProps.stepState?.isActive === nextProps.stepState?.isActive &&
    prevProps.stepState?.stepIndex === nextProps.stepState?.stepIndex &&
    prevProps.stepState?.isCompleted === nextProps.stepState?.isCompleted &&
    prevProps.stepState?.isLast === nextProps.stepState?.isLast
);
