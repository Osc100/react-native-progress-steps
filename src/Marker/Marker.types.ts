import type {
  ImarkerColors,
  IstepState,
  Torientation,
} from '../ProgressSteps.types';

export interface IMarker {
  colors?: ImarkerColors;
  orientation?: Torientation;
  stepState?: IstepState;
}

export type Tstyles = {
  [key: string]: {
    [key: string]: {
      [key: string]: string | number;
    };
  };
};
