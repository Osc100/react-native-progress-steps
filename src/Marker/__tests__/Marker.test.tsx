import React from 'react';
import Marker from '../';
import { render } from '@testing-library/react-native';

const stepState = {
  index: 0,
  isActive: true,
  isCompleted: false,
  isFirst: true,
  isLast: false,
  isFirstInteraction: true,
};

describe('Marker', () => {
  it('should render Marker', () => {
    const { queryByText, toJSON } = render(<Marker stepState={stepState} />);

    expect(queryByText('0')).toBeTruthy();
    expect(toJSON()).toMatchSnapshot();
  });
  it('should render last Marker', () => {
    const { toJSON } = render(
      <Marker stepState={{ ...stepState, isLast: true }} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
