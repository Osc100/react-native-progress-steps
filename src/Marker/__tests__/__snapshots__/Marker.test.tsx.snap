// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marker should render Marker 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#000000",
        "borderRadius": 20,
        "borderWidth": 4,
        "fontSize": 1,
        "height": 40,
        "justifyContent": "center",
        "position": "relative",
        "width": 40,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontWeight": "bold",
          "position": "relative",
          "zIndex": 1,
        }
      }
    >
      1
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 40,
        }
      }
    />
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "flex": 1,
          "position": "relative",
          "width": 4,
          "zIndex": 0,
        },
        Object {
          "backgroundColor": "#a4a4a4",
        },
        false,
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "height": "0%",
          "width": 4,
        }
      }
    />
  </View>
</View>
`;

exports[`Marker should render last Marker 1`] = `
<View
  onLayout={[Function]}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderColor": "#000000",
        "borderRadius": 20,
        "borderWidth": 4,
        "fontSize": 1,
        "height": 40,
        "justifyContent": "center",
        "position": "relative",
        "width": 40,
        "zIndex": 1,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#000000",
          "fontWeight": "bold",
          "position": "relative",
          "zIndex": 1,
        }
      }
    >
      1
    </Text>
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 20,
          "height": 40,
          "position": "absolute",
          "transform": Array [
            Object {
              "scale": 0,
            },
          ],
          "width": 40,
        }
      }
    />
  </View>
</View>
`;
